File: shared_resources/src/main/java/com/thewizrd/shared_resources/ApplicationLib.java
package com.thewizrd.shared_resources;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;

import androidx.annotation.NonNull;

import com.thewizrd.shared_resources.utils.SettingsManager;

public interface ApplicationLib {
    Context getAppContext();

    SharedPreferences getPreferences();

    void registerAppSharedPreferenceListener();

    void unregisterAppSharedPreferenceListener();

    void registerAppSharedPreferenceListener(@NonNull SharedPreferences.OnSharedPreferenceChangeListener listener);

    void unregisterAppSharedPreferenceListener(@NonNull SharedPreferences.OnSharedPreferenceChangeListener listener);

    AppState getAppState();

    boolean isPhone();

    Bundle getProperties();

    SettingsManager getSettingsManager();
}

File: shared_resources/src/main/java/com/thewizrd/shared_resources/ApplicationLib.kt
package com.thewizrd.shared_resources

import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import com.thewizrd.shared_resources.preferences.SettingsManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.MainScope

public lateinit var appLib: ApplicationLib

abstract class ApplicationLib {
    abstract val context: Context
    abstract val preferences: SharedPreferences
    abstract fun registerAppSharedPreferenceListener()
    abstract fun unregisterAppSharedPreferenceListener()
    abstract fun registerAppSharedPreferenceListener(listener: SharedPreferences.OnSharedPreferenceChangeListener)
    abstract fun unregisterAppSharedPreferenceListener(listener: SharedPreferences.OnSharedPreferenceChangeListener)
    abstract val appState: AppState
    abstract val isPhone: Boolean
    abstract val properties: Bundle
    abstract val settingsManager: SettingsManager
    open val appScope: CoroutineScope = MainScope()
}